torch/csrc/Module.cpp

```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
+    Py_RETURN_NONE;
+}
+
+// SIGCHLD hander should be registered on main loader process to catch any
+// worker failing. SIGALRM handler is needed for implementing timeout.
+// Python handles are _set_main_signal_handers() and
+// _remove_main_signal_handers().
+static void handler_SIGCHLD_main(int sig) {
+  int status;
+  pid_t p;
+  pid_t *pid_ptr;
+
+  // The flags and status checks ensure that we are really observing a child
+  // exiting, rather than other cases such as SIGSTOP and SIGCONT.
+  // https://stackoverflow.com/a/40707100
+  while ((p = waitpid(-1, &status, WNOHANG|WUNTRACED|WCONTINUED)) > 0) {
+    if (WIFCONTINUED(status) || WIFSTOPPED(status))
+      continue;
+    if (WIFEXITED(status) != 0 && WEXITSTATUS(status) == 0)
+      continue;
+    // child must have exited with signal/error, check if it is one of the pid
+    // we care about.
+    pid_ptr = worker_pids;
+    for (size_t i = 0; i < num_worker_pids; i++) {
+      if (*pid_ptr == p)
+        _exit(EXIT_FAILURE);
+      pid_ptr++;
+    }
+  }
+}
+
+SIGNAL_HANDLER(SIGALRM, handler_SIGALRM, "ERROR: Time out when fetching data in DataLoader.\n")
+
+// We don't want to exit on any SIGCHLD from any child. child_pids is a sequence
+// of pids we are interested in.
+PyObject *THPModule_setMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  auto tuple = PyTuple_Check(child_pids);
+  THPUtils_assert(tuple || PyList_Check(child_pids), "_set_main_signal_handler "
+        "expects a tuple or list, but got %s", THPUtils_typename(child_pids));
+
+  auto size = tuple ? PyTuple_GET_SIZE(child_pids) : PyList_GET_SIZE(child_pids);
+  {
+    std::unique_lock<std::mutex> lock(worker_pid_mutex);
+    for (int idx = 0; idx < size; idx++) {
+      PyObject* obj = tuple ? PyTuple_GET_ITEM(child_pids, idx) : PyList_GET_ITEM(child_pids, idx);
+      worker_pid_vec.push_back((pid_t) THPUtils_unpackLong(obj));
+    }
+    worker_pids = &worker_pid_vec[0];
+    num_worker_pids = worker_pid_vec.size();
+  }
+
+  signal(SIGCHLD, handler_SIGCHLD_main);
+  signal(SIGALRM, handler_SIGALRM);
+  Py_RETURN_NONE;
+}
+
+PyObject *THPModule_removeMainSignalHandlers(PyObject *module, PyObject *child_pids) {
+  auto tuple = PyTuple_Check(child_pids);
+  THPUtils_assert(tuple || PyList_Check(child_pids), "_remove_main_signal_handler "
+        "expects a tuple or list, but got %s", THPUtils_typename(child_pids));
+
+  auto size = tuple ? PyTuple_GET_SIZE(child_pids) : PyList_GET_SIZE(child_pids);
+  std::set<pid_t> pid_set;
+  for (int idx = 0; idx < size; idx++) {
+    PyObject* obj = tuple ? PyTuple_GET_ITEM(child_pids, idx) : PyList_GET_ITEM(child_pids, idx);
+    pid_set.insert((pid_t) THPUtils_unpackLong(obj));
+  }
+  {
+    std::unique_lock<std::mutex> lock(worker_pid_mutex);
+    worker_pid_vec.erase(std::remove_if(worker_pid_vec.begin(), worker_pid_vec.end(),
+      [&](pid_t p){return pid_set.find(p) != pid_set.end();}), worker_pid_vec.end());
+    worker_pids = &worker_pid_vec[0];
```

Comment:
I'm not sure how to fix this, but all this manipulation seems really unsafe. You're fiddling with the vector (making it smaller), even though `num_worker_pids` still contains the large value, so these elements at the end might still be read by the signal handler.